### ISSUE TITLE
Dynamic "Evidence" creation

### ISSUE THREAD

**Issue by benloh** [theRAPTLab/meme-2023#53]
Joshua writes: "It'd be nice to be able to "inject" evidence. This could be either by just dropping files into a location or by calling an API.  

Inject - yeah - like you wander around in their virtual world, and make a notebook with info. Then you go into meme to model what you learned, and in meme you see the same information from the virtual world to help guide your modeling.

Right now, the approach would be: Joshua goes into virtual world, exports needed things, makes them into files, drops in resources folder, then goes to admin and adds.

If the future iteration allowed a simple way to either drop the files in with info directly, or call an API call to create on the fly ... that'd be cool."

* Use filenames as labels
* Add numeric naming convention as reference numbers, e.g. "12-Resource Table 4.pdf".
* Associating evidence with a classroom would still require admin UI

**Comment by benloh** [comment]
@jdanish Does dynamic injection happen in the middle of a student session?  Or is this only done during, say curriculum unit development before or after school?

**Comment by jdanish** [comment]
When I called it that I was thinking middle of unit but for now doing it outside of lessons by the teacher would be amazing.

**Comment by benloh** [comment]
Thanks.  It does have a significant impact on the architecture.  But if we don't have to worry about live updates in the middle of a student session, it can be simpler.  We can enhance it with the next grant.

**Comment by jdanish** [comment]
Yeah. At the time we were hoping the AI app (different project) could inject things but I agree, later grant. Doing it after hours by the teacher would be lovely. Or saying "Everyone log out for a second" in extreme cases.

**Comment by jdanish** [comment]
As a note, you can currently show and hide evidence in the middle of a session, and that works well.  I realize those are different functions and this is likely much harder - just noting it so we should have a way of saying either "don't do this now" or "this won't show up until later" etc.

**Comment by benloh** [comment]
> As a note, you can currently show and hide evidence in the middle of a session, and that works well. I realize those are different functions and this is likely much harder - just noting it so we should have a way of saying either "don't do this now" or "this won't show up until later" etc.

Hmm...so you still need to be able to show or hide a resource in the middle of a session, rather than simply enabling all the resources for a unit?  This does get rather complicated.

Is this because you're gradually revealing sources when students ask for them?  And then the state of enable/disabled needs to be maintained for that unit over time?  At this point we are no longer simply reading a static unit definition but need to dynamically update the unit definitions.

**Comment by jdanish** [comment]
Not really - it just happened to work that way and I think the only time we used it was when we forgot to turn something on. Long-term, yes it'd be nice if the teacher could "reveal" new evidence mid-class. Given our time constraints, though, it is fine if new evidence can be added and activated only after / before class. Though to be clear, since we have no way currently of seeing that students are logged in, I assume that means that students will only see new evidence in their first login after we make changes?  Which is AOK. In typical class contexts, the kids are only online during the actual class session, so we can assume they are not logged in when we make edits that night. So if the next time they login they see whatever is "current" that's fine. If they need to have officially logged off, that's also fine since I believe there is a timeout, but we'd want to make sure there is a "force logoff" option in admin in case something funny happened at the end of class and a logout didn't register?

**Comment by benloh** [comment]
Yeah, we're going to need to review all of this.  A few notes:
- [ ] Add "Force Logoff" to force students to reload?
- [ ] Add "Reload Server" to force resource reload after changes to PDFs or Unit definitions?

**Comment by benloh** [comment]
Another complicating factor: Electron support for this is iffy at best.  It probably won't work.
We're probably going to end up in a bad technical debt situation ala Template updates with Net.Create.

- [ ] Review Electron management of resources.  How do we manage packaging?

**Comment by jdanish** [comment]
I realize the notes are for you, but some quick notes:

- Currently, the system appears to draw from resources, and / or screenshots as-needed
   - If you click on the checkbox in admin to make a resource visible, it immediately becomes visible in all students who are connected
   - if you click in student a to attach a screenshot to an evidence link, student b can see it immediately by opening that link
   - If I recall, electron will create a resource folder / data folder and then things can be dropped in there. We'd be fine with continuing to copy the resources and / or data by hand for electron distribution to avoid added challenge / complexity

**Comment by benloh** [comment]
Implemented with Units.  #137
